### PR TITLE
Config Structural Validation Schema Part 2: Security and Policy

### PR DESCRIPTION
**Pull request by Zash**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [ ] personal data beyond what is necessary for interacting with this pull request, nor
> - [ ] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [x] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

This adds structural validation for the following top-level config sections:

- clusterAdmin
- dex
- externalTrafficPolicy
- falco
- gatekeeper
- kured
- opa
- trivy

For more context including the required scripts and see #1862
Part of issue #1427

#### Information to reviewers

`ck8s validate` should now perform validation of the above sections.

While the CI is failing:
Please help determine whether it is the schema that is lacking or whether it detected a config mistake.

Suggestions and improvements very welcome!

[Also see the **README** for brief information about the schema format](https://github.com/elastisys/compliantkubernetes-apps/blob/main/config/schemas/README.md)

**Obs:** The `$defs` are really added in #2063 but the commit is included here too for CI.

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [x] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
  - [x] The change updates the config *and* the schema
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests

### CODE REVIEW

**Comment by Zash** [comment]
I will rebase this once #2063 has been merged.

File: config/schemas/config.yaml [hunk 1]
```diff
@@ -1358,6 +1358,612 @@ properties:
...
+                - modern-bpf
+            ebpf:
+              additionalProperties: false
+              properties:
+                hostNetwork:
+                  description: Needed to enable eBPF JIT at runtime for performance reasons. Can be skipped if eBPF JIT is enabled from outside the container
+                  default: false
+                  type: boolean
+                path:
+                  description: Path where the eBPF probe is located. It comes handy when the probe have been installed in the nodes using tools other than the init container deployed with the chart.
```
Comment (OlleLarsson) [inline config/schemas/config.yaml:1568]:
```suggestion
                  description: Path where the eBPF probe is located. It comes in handy when the probe have been installed in the nodes using tools other than the init container deployed with the chart.
```

**Review by OlleLarsson — APPROVED** [review]
Nice!

OK from what I tested. Didn't check everything but tried some invalid configuration to see that it complained.

Q: Should it return something other than 0 is the validation fails :) ?

```
./bin/ck8s validate wc; echo $?
...
[ck8s] Do you want to abort? (y/N): n
[ck8s] Failed schema validation:
wc-config.yaml: clusterAdmin.groups: Invalid type. Expected: array, given: object
1 of 1 failed validation
.clusterAdmin.groups = {}
[ck8s] Do you want to abort? (y/N): n
Config validation successful
0
```

edit: ^ seems to be the config validation and not the schema validation that succeeds, I got confused.

File: config/schemas/config.yaml [hunk 2]
```diff
@@ -1358,6 +1358,612 @@ properties:
...
+        type: boolean
+    type: object
+  gatekeeper: true
+  kured:
+    additionalProperties: false
+    properties:
+      affinity:
+        $ref: '#/$defs/io.k8s.api.core.v1.Affinity'
+      configuration:
+        $comment: Would it be possible to extract schema from somewhere? Guess from the thing maybe?
```
Comment (Pavan-Gunda) [inline config/schemas/config.yaml:1652]:
```suggestion
        $comment: Would it be possible to extract schema from somewhere else in the future? 
```

Should we make this a little clearer so the next person working on it knows this is a thought left for future work.

**Comment by OlleLarsson** [comment]
> Nice!
> 
> OK from what I tested. Didn't check everything but tried some invalid configuration to see that it complained.
> 
> Q: Should it return something other than 0 is the validation fails :) ?
> 
> ```
> ./bin/ck8s validate wc; echo $?
> ...
> [ck8s] Do you want to abort? (y/N): n
> [ck8s] Failed schema validation:
> wc-config.yaml: clusterAdmin.groups: Invalid type. Expected: array, given: object
> 1 of 1 failed validation
> .clusterAdmin.groups = {}
> [ck8s] Do you want to abort? (y/N): n
> Config validation successful
> 0
> ```



> Nice!
> 
> OK from what I tested. Didn't check everything but tried some invalid configuration to see that it complained.
> 
> Q: Should it return something other than 0 is the validation fails :) ?
> 
> ```
> ./bin/ck8s validate wc; echo $?
> ...
> [ck8s] Do you want to abort? (y/N): n
> [ck8s] Failed schema validation:
> wc-config.yaml: clusterAdmin.groups: Invalid type. Expected: array, given: object
> 1 of 1 failed validation
> .clusterAdmin.groups = {}
> [ck8s] Do you want to abort? (y/N): n
> Config validation successful
> 0
> ```

edit: ^ seems to be the config validation that succeeds and not the schema validation, i got confused.

**Review by lucianvlad — APPROVED** [review]
LGTM

With comments from Olle and Pavan

**Comment by Zash** [comment]
> Q: Should it return something other than 0 is the validation fails :) ?

I believe it behaved this way even before #1862 .
It exists with non-zero if you say yes to aborting, otherwise zero.

**Review by Pavan-Gunda — APPROVED** [review]

**Comment by Zash** [comment]
Last-minute change: Added `gatekeeper` which was previously unspecified.

**Pull request merged**